src/wamr/WAMRWasmModule.cpp

```diff
@@ -187,29 +199,139 @@ int WAMRWasmModule::executeWasmFunction(const std::string& funcName)
     } else {
         std::string errorMessage(
           ((AOTModuleInstance*)moduleInstance)->cur_exception);
-        SPDLOG_ERROR("Function failed: {}", errorMessage);
+        SPDLOG_ERROR("Caught wasm runtime exception: {}", errorMessage);
+
+        // TODO - agree on a protocol to send the return code through the error
+        // message (throwing exceptions not supported). For the moment, if the
+        // error message is a "Exception: 0" we assume no errors (i.e. exit(0))
+        if (errorMessage == "Exception: 0") {
+            return 0;
+        }
 
         return 1;
     }
 
     return 0;
 }
 
+void WAMRWasmModule::writeStringToWasmMemory(const std::string& strHost,
+                                             char* strWasm)
+{
+    this->validateNativeAddress(strWasm, strHost.size());
+    std::copy(strHost.begin(), strHost.end(), strWasm);
+}
+
+void WAMRWasmModule::writeStringArrayToMemory(
+  const std::vector<std::string>& strings,
+  uint32_t* strOffsets,
+  char* strBuffer)
+{
+    this->validateNativeAddress(strOffsets, strings.size() * sizeof(uint32_t));
+
+    char* nextBuffer = strBuffer;
+    size_t i;
+    for (i = 0; i < strings.size(); ++i) {
+        const std::string& thisStr = strings.at(i);
+
+        this->validateNativeAddress(nextBuffer, thisStr.size() + 1);
+
+        std::copy(thisStr.begin(), thisStr.end(), nextBuffer);
+        strOffsets[i] = this->nativePointerToWasm(nextBuffer);
+
+        nextBuffer += thisStr.size() + 1;
+    }
+}
+
+void WAMRWasmModule::writeArgvToWamrMemory(uint32_t* argvOffsetsWasm,
+                                           char* argvBuffWasm)
+{
+    writeStringArrayToMemory(argv, argvOffsetsWasm, argvBuffWasm);
+}
+
+void WAMRWasmModule::writeWasmEnvToWamrMemory(uint32_t* envOffsetsWasm,
+                                              char* envBuffWasm)
+{
+    writeStringArrayToMemory(
+      wasmEnvironment.getVars(), envOffsetsWasm, envBuffWasm);
+}
+
+void WAMRWasmModule::validateNativeAddress(void* nativePtr, size_t size)
+{
+    if (!wasm_runtime_validate_native_addr(moduleInstance, nativePtr, size)) {
+        SPDLOG_ERROR("Native pointer does not point to a valid address in the "
+                     "WAMR's module address space");
+        throw std::runtime_error("Pointer out of WAMR's module address space");
+    }
+}
+
+void WAMRWasmModule::validateWasmOffset(uint32_t wasmOffset, size_t size)
+{
+    if (!wasm_runtime_validate_app_addr(moduleInstance, wasmOffset, size)) {
+        SPDLOG_ERROR("WASM offset outside WAMR module instance memory");
+        throw std::runtime_error("Offset outside WAMR's memory");
+    }
+}
+
+uint8_t* WAMRWasmModule::wasmPointerToNative(uint32_t wasmPtr)
+{
+    void* nativePtr = wasm_runtime_addr_app_to_native(moduleInstance, wasmPtr);
+    if (nativePtr == nullptr) {
+        SPDLOG_ERROR(
+          "WASM offset {} is out of the WAMR module's address space");
```

Comment:
This has a space for a variable in the message but is missing that variable so won't print anything at runtime.